protocol/opcm-bytecode-verification.md

```diff
@@ -0,0 +1,168 @@
+# OPCM Bytecode Verification: Design Doc
+
+|                    |            |
+| ------------------ | ---------- |
+| Author             | Maurelian  |
+| Created at         | 2025-06-04 |
+| Initial Reviewers  | TBD        |
+| Need Approval From | TBD        |
+| Status             | Draft      |
+
+## Purpose
+
+Ensure that the contracts referenced in `OPContractsManager` (OPCM) are verifiably built from
+trusted source code by introducing a bytecode verification step into the release process.
+
+## Summary
+
+We propose integrating the `VerifyOPCM.s.sol` Foundry script into the release process via
+`op-deployer` to validate that deployed contract bytecode matches locally built artifacts. This
+prevents accidental or malicious mismatches and establishes trust in deployments originating from a
+known commit. The step will be automated in CI, and become part of the documented SDLC.
+
+## Problem Statement + Context
+
+Currently, there is no enforced mechanism to ensure that the OPCM used in an upgrade is built from a
+trusted commit, which should be one labelled as an `op-contracts/vX.Y.Z` tag, and approved by
+governance.
+
+This creates a risk of human error or maliciousness leading to an upgrade performed by an incorrect
+version of the OPCM (or the implementation contracts it sets). It also presents a risk of a failed
+upgrade resulting from a misconfigured OPCM (ie. if any [constructor
+vars](https://github.com/ethereum-optimism/optimism/blob/a10fd5259a3af9a465955b035e16f516327d51d5/packages/contracts-bedrock/src/L1/OPContractsManager.sol#L266-L269)
+are set incorrectly).
```

Comment:
We need a strategy for ensuring that VerifyOPCM remains complete, so that if a new contract is added to the system, that will be identified.